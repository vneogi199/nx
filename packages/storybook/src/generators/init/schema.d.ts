import { UiFramework } from '../../utils/models';

export interface Schema {
  uiFramework: UiFramework;
  js?: boolean;
}
