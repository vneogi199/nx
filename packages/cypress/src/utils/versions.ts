export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.13.4';
export const typesNodeVersion = '16.11.7';
export const cypressViteDevServerVersion = '^2.2.1';
export const cypressVersion = '^13.0.0';
export const cypressWebpackVersion = '^2.0.0';
export const webpackHttpPluginVersion = '^5.5.0';
export const viteVersion = '~4.3.9';
export const htmlWebpackPluginVersion = '^5.5.0';
