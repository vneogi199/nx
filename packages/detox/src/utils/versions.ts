export const nxVersion = require('../../package.json').version;

export const detoxVersion = '^20.11.1';
export const testingLibraryJestDom = '5.16.5';
export const configPluginsDetoxVersion = '~6.0.0'; // only required for expo
