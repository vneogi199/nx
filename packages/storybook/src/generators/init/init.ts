import {
  addDependenciesToPackageJson,
  detectPackageManager,
  GeneratorCallback,
  readJson,
  readNxJson,
  runTasksInSerial,
  Tree,
  updateJson,
  updateNxJson,
} from '@nx/devkit';
import { initGenerator as jsInitGenerator } from '@nx/js';

import {
  litVersion,
  nxVersion,
  reactVersion,
  storybookVersion,
  storybookReactNativeVersion,
  viteVersion,
} from '../../utils/versions';
import { Schema } from './schema';
import {
  getInstalledStorybookVersion,
  storybookMajorVersion,
} from '../../utils/utilities';
import { gte } from 'semver';

function checkDependenciesInstalled(host: Tree, schema: Schema) {
  const packageJson = readJson(host, 'package.json');

  const devDependencies = {};
  const dependencies = {};
  packageJson.dependencies = packageJson.dependencies || {};
  packageJson.devDependencices = packageJson.devDependencices || {};

  // base deps
  devDependencies['@nx/storybook'] = nxVersion;

  let storybook7VersionToInstall = storybookVersion;
  if (
    storybookMajorVersion() === 7 &&
    getInstalledStorybookVersion() &&
    gte(getInstalledStorybookVersion(), '7.0.0')
  ) {
    storybook7VersionToInstall = getInstalledStorybookVersion();
  }

  // Needed for Storybook 7
  // https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react-peer-dependencies-required
  if (
    !packageJson.dependencies['react'] &&
    !packageJson.devDependencies['react']
  ) {
    dependencies['react'] = reactVersion;
  }
  if (
    !packageJson.dependencies['react-dom'] &&
    !packageJson.devDependencies['react-dom']
  ) {
    dependencies['react-dom'] = reactVersion;
  }

  devDependencies['@storybook/core-server'] = storybook7VersionToInstall;
  devDependencies['@storybook/addon-essentials'] = storybook7VersionToInstall;

  if (schema.uiFramework) {
    if (schema.uiFramework === '@storybook/react-native') {
      devDependencies['@storybook/react-native'] = storybookReactNativeVersion;
    } else {
      devDependencies[schema.uiFramework] = storybook7VersionToInstall;
      const isPnpm = detectPackageManager(host.root) === 'pnpm';
      if (isPnpm) {
        // If it's pnpm, it needs the framework without the builder
        // as a dependency too (eg. @storybook/react)
        const matchResult = schema.uiFramework?.match(/^@storybook\/(\w+)/);
        const uiFrameworkWithoutBuilder = matchResult ? matchResult[0] : null;
        if (uiFrameworkWithoutBuilder) {
          devDependencies[uiFrameworkWithoutBuilder] =
            storybook7VersionToInstall;
        }
      }
    }

    if (schema.uiFramework === '@storybook/vue3-vite') {
      if (
        !packageJson.dependencies['@storybook/vue3'] &&
        !packageJson.devDependencies['@storybook/vue3']
      ) {
        devDependencies['@storybook/vue3'] = storybook7VersionToInstall;
      }
    }

    if (schema.uiFramework === '@storybook/angular') {
      if (
        !packageJson.dependencies['@angular/forms'] &&
        !packageJson.devDependencies['@angular/forms']
      ) {
        devDependencies['@angular/forms'] = '*';
      }
    }

    if (
      schema.uiFramework === '@storybook/web-components-vite' ||
      schema.uiFramework === '@storybook/web-components-webpack5'
    ) {
      devDependencies['lit'] = litVersion;
    }

    if (schema.uiFramework === '@storybook/react-native') {
      devDependencies['@storybook/addon-ondevice-actions'] =
        storybookReactNativeVersion;
      devDependencies['@storybook/addon-ondevice-backgrounds'] =
        storybookReactNativeVersion;
      devDependencies['@storybook/addon-ondevice-controls'] =
        storybookReactNativeVersion;
      devDependencies['@storybook/addon-ondevice-notes'] =
        storybookReactNativeVersion;
    }

    if (schema.uiFramework.endsWith('-vite')) {
      if (
        !packageJson.dependencies['vite'] &&
        !packageJson.devDependencies['vite']
      ) {
        devDependencies['vite'] = viteVersion;
      }
    }
  }

  return addDependenciesToPackageJson(host, dependencies, devDependencies);
}

function addCacheableOperation(tree: Tree) {
  const nxJson = readNxJson(tree);
  const cacheableOperations: string[] | null =
    nxJson.tasksRunnerOptions?.default?.options?.cacheableOperations;

  if (cacheableOperations && cacheableOperations.includes('build-storybook')) {
    nxJson.tasksRunnerOptions.default.options.cacheableOperations.push(
      'build-storybook'
    );
  }

  nxJson.targetDefaults ??= {};
  nxJson.targetDefaults['build-storybook'] ??= {};
  nxJson.targetDefaults['build-storybook'].cache = true;

  updateNxJson(tree, nxJson);
}

function moveToDevDependencies(tree: Tree) {
  updateJson(tree, 'package.json', (packageJson) => {
    packageJson.dependencies = packageJson.dependencies || {};
    packageJson.devDependencies = packageJson.devDependencies || {};

    if (packageJson.dependencies['@nx/storybook']) {
      packageJson.devDependencies['@nx/storybook'] =
        packageJson.dependencies['@nx/storybook'];
      delete packageJson.dependencies['@nx/storybook'];
    }
    return packageJson;
  });
}

/**
 * This is a temporary fix for Storybook to support TypeScript configuration files.
 * The issue is that if there is a root tsconfig.json file, Storybook will use it, and
 * ignore the tsconfig.json file in the .storybook folder. This results in module being set
 * to esnext, and Storybook does not recognise the main.ts code as a module.
 */
function editRootTsConfig(tree: Tree) {
  if (tree.exists('tsconfig.json')) {
    updateJson(tree, 'tsconfig.json', (json) => {
      if (json['ts-node']) {
        json['ts-node'] = {
          ...json['ts-node'],
          compilerOptions: {
            ...(json['ts-node'].compilerOptions ?? {}),
            module: 'commonjs',
          },
        };
      } else {
        json['ts-node'] = {
          compilerOptions: {
            module: 'commonjs',
          },
        };
      }
      return json;
    });
  }
}

export async function initGenerator(tree: Tree, schema: Schema) {
  const tasks: GeneratorCallback[] = [];
  tasks.push(
    await jsInitGenerator(tree, {
      ...schema,
      skipFormat: true,
    })
  );
  tasks.push(checkDependenciesInstalled(tree, schema));
  moveToDevDependencies(tree);
  editRootTsConfig(tree);
  addCacheableOperation(tree);
  return runTasksInSerial(...tasks);
}

export default initGenerator;
