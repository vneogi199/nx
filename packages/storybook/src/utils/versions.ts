export const nxVersion = require('../../package.json').version;
export const storybookReactNativeVersion = '^6.5.3';
export const reactNativeStorybookLoader = '^2.0.5';
export const storybookTestRunnerVersion = '^0.13.0';
export const storybookTestingLibraryVersion = '~0.2.0';
export const storybookJestVersion = '~0.1.0';
export const litVersion = '^2.6.1';
export const tsNodeVersion = '10.9.1';

export const storybookVersion = '^7.5.1';
export const reactVersion = '^18.2.0';
export const viteVersion = '~4.3.9';

export const coreJsVersion = '^3.6.5';
