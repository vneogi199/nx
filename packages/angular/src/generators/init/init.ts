import {
  addDependenciesToPackageJson,
  ensurePackage,
  formatFiles,
  GeneratorCallback,
  logger,
  readNxJson,
  runTasksInSerial,
  Tree,
  updateNxJson,
} from '@nx/devkit';
import { jestInitGenerator } from '@nx/jest';
import { Linter } from '@nx/eslint';
import { initGenerator as jsInitGenerator } from '@nx/js';
import { E2eTestRunner, UnitTestRunner } from '../../utils/test-runners';
import {
  addDependenciesToPackageJsonIfDontExist,
  getInstalledPackageVersion,
  versions,
} from '../utils/version-utils';
import type { PackageVersions } from '../../utils/backward-compatible-versions';
import { Schema } from './schema';
import { nxVersion } from '../../utils/versions';

export async function angularInitGenerator(
  tree: Tree,
  rawOptions: Schema
): Promise<GeneratorCallback> {
  const tasks: GeneratorCallback[] = [];
  const options = normalizeOptions(rawOptions);

  const pkgVersions = versions(tree);

  const peerDepsToInstall = ['@angular-devkit/core'];
  let devkitVersion: string;
  peerDepsToInstall.forEach((pkg) => {
    const packageVersion = getInstalledPackageVersion(tree, pkg);

    if (!packageVersion) {
      devkitVersion ??=
        getInstalledPackageVersion(tree, '@angular-devkit/build-angular') ??
        pkgVersions.angularDevkitVersion;

      try {
        ensurePackage(pkg, devkitVersion);
      } catch {
        // @schematics/angular cannot be required so this fails but this will still allow wrapping the schematic later on
      }

      if (!options.skipPackageJson) {
        tasks.push(
          addDependenciesToPackageJson(tree, {}, { [pkg]: devkitVersion })
        );
      }
    }
  });
  setDefaults(tree, options);

  const jsTask = await jsInitGenerator(tree, {
    ...options,
    tsConfigName: options.rootProject ? 'tsconfig.json' : 'tsconfig.base.json',
    js: false,
    skipFormat: true,
  });
  tasks.push(jsTask);

  if (!options.skipPackageJson) {
    tasks.push(updateDependencies(tree, pkgVersions));
  }
  const unitTestTask = await addUnitTestRunner(
    tree,
    options,
    pkgVersions.jestPresetAngularVersion
  );
  tasks.push(unitTestTask);
  const e2eTask = await addE2ETestRunner(tree, options);
  tasks.push(e2eTask);

  ignoreAngularCacheDirectory(tree);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return runTasksInSerial(...tasks);
}

function normalizeOptions(options: Schema): Required<Schema> {
  return {
    e2eTestRunner: options.e2eTestRunner ?? E2eTestRunner.Cypress,
    linter: options.linter ?? Linter.EsLint,
    skipFormat: options.skipFormat ?? false,
    skipInstall: options.skipInstall ?? false,
    skipPackageJson: options.skipPackageJson ?? false,
    style: options.style ?? 'css',
    unitTestRunner: options.unitTestRunner ?? UnitTestRunner.Jest,
    rootProject: options.rootProject,
  };
}

function setDefaults(host: Tree, options: Schema) {
  const nxJson = readNxJson(host);

  nxJson.generators = nxJson.generators || {};
  nxJson.generators['@nx/angular:application'] = {
    style: options.style,
    linter: options.linter,
    unitTestRunner: options.unitTestRunner,
    e2eTestRunner: options.e2eTestRunner,
    ...(nxJson.generators['@nx/angular:application'] || {}),
  };
  nxJson.generators['@nx/angular:library'] = {
    linter: options.linter,
    unitTestRunner: options.unitTestRunner,
    ...(nxJson.generators['@nx/angular:library'] || {}),
  };
  nxJson.generators['@nx/angular:component'] = {
    style: options.style,
    ...(nxJson.generators['@nx/angular:component'] || {}),
  };

  updateNxJson(host, nxJson);
}

function updateDependencies(
  tree: Tree,
  versions: PackageVersions
): GeneratorCallback {
  const angularVersion =
    getInstalledPackageVersion(tree, '@angular/core') ??
    versions.angularVersion;
  const angularDevkitVersion =
    getInstalledPackageVersion(tree, '@angular-devkit/build-angular') ??
    versions.angularDevkitVersion;
  const rxjsVersion =
    getInstalledPackageVersion(tree, 'rxjs') ?? versions.rxjsVersion;
  const tsLibVersion =
    getInstalledPackageVersion(tree, 'tslib') ?? versions.tsLibVersion;
  const zoneJsVersion =
    getInstalledPackageVersion(tree, 'zone.js') ?? versions.zoneJsVersion;

  return addDependenciesToPackageJsonIfDontExist(
    tree,
    {
      '@angular/animations': angularVersion,
      '@angular/common': angularVersion,
      '@angular/compiler': angularVersion,
      '@angular/core': angularVersion,
      '@angular/forms': angularVersion,
      '@angular/platform-browser': angularVersion,
      '@angular/platform-browser-dynamic': angularVersion,
      '@angular/router': angularVersion,
      rxjs: rxjsVersion,
      tslib: tsLibVersion,
      'zone.js': zoneJsVersion,
    },
    {
      '@angular/cli': angularDevkitVersion,
      '@angular/compiler-cli': angularVersion,
      '@angular/language-service': angularVersion,
      '@angular-devkit/build-angular': angularDevkitVersion,
      '@angular-devkit/schematics': angularDevkitVersion,
      '@schematics/angular': angularDevkitVersion,
    }
  );
}

async function addUnitTestRunner(
  tree: Tree,
  options: Schema,
  jestPresetAngularVersion: string
): Promise<GeneratorCallback> {
  switch (options.unitTestRunner) {
    case UnitTestRunner.Jest:
      if (!options.skipPackageJson) {
        process.env.npm_config_legacy_peer_deps ??= 'true';

        addDependenciesToPackageJsonIfDontExist(
          tree,
          {},
          {
            'jest-preset-angular': jestPresetAngularVersion,
          }
        );
      }

      return jestInitGenerator(tree, {
        skipPackageJson: options.skipPackageJson,
      });
    default:
      return () => {};
  }
}

async function addE2ETestRunner(
  tree: Tree,
  options: Schema
): Promise<GeneratorCallback> {
  switch (options.e2eTestRunner) {
    case E2eTestRunner.Cypress:
      const { cypressInitGenerator } = ensurePackage<
        typeof import('@nx/cypress')
      >('@nx/cypress', nxVersion);
      return cypressInitGenerator(tree, {
        skipPackageJson: options.skipPackageJson,
      });
    case E2eTestRunner.Playwright:
      const { initGenerator: playwrightInitGenerator } = ensurePackage<
        typeof import('@nx/playwright')
      >('@nx/playwright', nxVersion);
      return playwrightInitGenerator(tree, {
        skipFormat: true,
        skipPackageJson: options.skipPackageJson,
      });
    default:
      return () => {};
  }
}

function ignoreAngularCacheDirectory(tree: Tree): void {
  const { cli } = readNxJson(tree);
  // angular-specific cli config is supported though is not included in the
  // NxJsonConfiguration type
  const angularCacheDir = (cli as any)?.cache?.path ?? '.angular';

  addGitIgnoreEntry(tree, angularCacheDir);
  addPrettierIgnoreEntry(tree, angularCacheDir);
}

function addGitIgnoreEntry(tree: Tree, entry: string): void {
  if (tree.exists('.gitignore')) {
    let content = tree.read('.gitignore', 'utf-8');
    if (/^\.angular$/gm.test(content)) {
      return;
    }

    content = `${content}\n${entry}\n`;
    tree.write('.gitignore', content);
  } else {
    logger.warn(`Couldn't find .gitignore file to update`);
  }
}

function addPrettierIgnoreEntry(tree: Tree, entry: string): void {
  if (!tree.exists('.prettierignore')) {
    return;
  }

  let content = tree.read('.prettierignore', 'utf-8');
  if (/^\.angular(\/cache)?$/gm.test(content)) {
    return;
  }

  content = `${content}\n${entry}\n`;
  tree.write('.prettierignore', content);
}

export default angularInitGenerator;
