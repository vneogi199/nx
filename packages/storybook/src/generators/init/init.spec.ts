import {
  addDependenciesToPackageJson,
  NxJsonConfiguration,
  readJson,
  Tree,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import { nxVersion } from '../../utils/versions';
import { initGenerator } from './init';

describe('@nx/storybook:init', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  describe('dependencies for package.json', () => {
    it('should add angular related dependencies when using Angular as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/angular',
      });
      const packageJson = readJson(tree, 'package.json');

      expect(packageJson).toMatchSnapshot();

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/core-server']
      ).toBeDefined();

      // angular specific
      expect(packageJson.devDependencies['@storybook/angular']).toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();
    });

    it('should add react related dependencies when using React as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/react-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();
    });
    it('should add html related dependencies when using html as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/html-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();
    });

    it('should add web-components related dependencies when using html as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/web-components-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic web-components specific
      expect(
        packageJson.devDependencies['@storybook/web-components-webpack5']
      ).toBeDefined();

      // generic vue specific
      expect(
        packageJson.devDependencies['@storybook/vue-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();
    });

    it('should add vue related dependencies when using vue as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/vue-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic web-components specific
      expect(
        packageJson.devDependencies['@storybook/web-components-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();

      // generic vue specific
      expect(
        packageJson.devDependencies['@storybook/vue-webpack5']
      ).toBeDefined();
    });

    it('should add vue3 related dependencies when using vue3 as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/vue3-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic vue specific
      expect(
        packageJson.devDependencies['@storybook/vue-webpack5']
      ).not.toBeDefined();

      // generic web-components specific
      expect(
        packageJson.devDependencies['@storybook/web-components-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).not.toBeDefined();

      // generic vue3 specific
      expect(
        packageJson.devDependencies['@storybook/vue3-webpack5']
      ).toBeDefined();
    });

    it('should add svelte related dependencies when using svelte as uiFramework', async () => {
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nx/storybook': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/svelte-webpack5',
      });
      const packageJson = readJson(tree, 'package.json');

      // general deps
      expect(packageJson.devDependencies['@nx/storybook']).toBeDefined();
      expect(packageJson.dependencies['@nx/storybook']).toBeUndefined();
      expect(packageJson.dependencies[existing]).toBeDefined();
      expect(packageJson.devDependencies[existing]).toBeDefined();
      expect(
        packageJson.devDependencies['@storybook/addon-essentials']
      ).toBeDefined();

      // react specific
      expect(
        packageJson.devDependencies['@storybook/react-webpack5']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@babel/core']).not.toBeDefined();
      expect(packageJson.devDependencies['babel-loader']).not.toBeDefined();

      // angular specific
      expect(
        packageJson.devDependencies['@storybook/angular']
      ).not.toBeDefined();
      expect(packageJson.devDependencies['@angular/forms']).not.toBeDefined();

      // generic html specific
      expect(
        packageJson.devDependencies['@storybook/html-webpack5']
      ).not.toBeDefined();

      // generic vue specific
      expect(
        packageJson.devDependencies['@storybook/vue-webpack5']
      ).not.toBeDefined();

      // generic web-components specific
      expect(
        packageJson.devDependencies['@storybook/web-components-webpack5']
      ).not.toBeDefined();

      // generic vue3 specific
      expect(
        packageJson.devDependencies['@storybook/vue3-webpack5']
      ).not.toBeDefined();

      // generic svelte specific
      expect(
        packageJson.devDependencies['@storybook/svelte-webpack5']
      ).toBeDefined();
    });
  });

  describe('cacheable operations', () => {
    it('should add build-storybook to cacheable operations', async () => {
      await initGenerator(tree, {
        uiFramework: '@storybook/html-webpack5',
      });
      const nxJson = readJson<NxJsonConfiguration>(tree, 'nx.json');
      expect(nxJson.targetDefaults['build-storybook'].cache).toEqual(true);
    });

    it('should add build-storybook to cacheable operations for web-components', async () => {
      await initGenerator(tree, {
        uiFramework: '@storybook/web-components-webpack5',
      });
      const nxJson = readJson<NxJsonConfiguration>(tree, 'nx.json');
      expect(nxJson.targetDefaults['build-storybook'].cache).toEqual(true);
    });

    it('should add build-storybook to cacheable operations for vue', async () => {
      await initGenerator(tree, {
        uiFramework: '@storybook/vue-webpack5',
      });
      const nxJson = readJson<NxJsonConfiguration>(tree, 'nx.json');
      expect(nxJson.targetDefaults['build-storybook'].cache).toEqual(true);
    });

    it('should add build-storybook to cacheable operations for vue3', async () => {
      await initGenerator(tree, {
        uiFramework: '@storybook/vue3-webpack5',
      });
      const nxJson = readJson(tree, 'nx.json');
      expect(nxJson.targetDefaults['build-storybook'].cache).toEqual(true);
    });

    it('should not add any framework specific dependencies when no framework is specified', async () => {
      await initGenerator(tree, { uiFramework: undefined });

      // get the updated package.json
      const packageJson = readJson(tree, 'package.json');

      // check that only the following dependencies have been added
      expect(Object.keys(packageJson.devDependencies)).toEqual([
        '@nx/js',
        '@nx/storybook',
        '@storybook/addon-essentials',
        '@storybook/core-server',
        '@swc-node/register',
        '@swc/core',
        'prettier',
        'typescript',
      ]);
    });
  });

  describe('update root tsconfig.json', () => {
    it('should set the tsnode module to commonjs if there is a root tsconfig.json', async () => {
      tree.write(
        'tsconfig.json',
        JSON.stringify({
          extends: './tsconfig.base.json',
          compilerOptions: {
            jsx: 'react-jsx',
            allowJs: true,
            esModuleInterop: true,
            allowSyntheticDefaultImports: true,
            forceConsistentCasingInFileNames: true,
            strict: true,
            noImplicitOverride: true,
            noPropertyAccessFromIndexSignature: true,
            noImplicitReturns: true,
            noFallthroughCasesInSwitch: true,
          },
          files: [],
          include: [],
          references: [
            {
              path: './tsconfig.app.json',
            },
            {
              path: './.storybook/tsconfig.json',
            },
          ],
        })
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/react-webpack5',
      });
      const tsconfig = readJson(tree, 'tsconfig.json');
      expect(tsconfig['ts-node'].compilerOptions.module).toEqual('commonjs');
    });

    it('should set the tsnode module to commonjs and respect other tsnode settings', async () => {
      tree.write(
        'tsconfig.json',
        JSON.stringify({
          extends: './tsconfig.base.json',
          compilerOptions: {
            jsx: 'react-jsx',
            allowJs: true,
            esModuleInterop: true,
            allowSyntheticDefaultImports: true,
            forceConsistentCasingInFileNames: true,
            strict: true,
            noImplicitOverride: true,
            noPropertyAccessFromIndexSignature: true,
            noImplicitReturns: true,
            noFallthroughCasesInSwitch: true,
          },
          files: [],
          include: [],
          'ts-node': {
            otherSetting: 'value',
            compilerOptions: {
              test: 'my-test-value',
              module: 'my-module-value',
            },
          },
        })
      );
      await initGenerator(tree, {
        uiFramework: '@storybook/react-webpack5',
      });
      const tsconfig = readJson(tree, 'tsconfig.json');
      expect(tsconfig['ts-node'].otherSetting).toEqual('value');
      expect(tsconfig['ts-node'].compilerOptions.module).toEqual('commonjs');
      expect(tsconfig['ts-node'].compilerOptions.test).toEqual('my-test-value');
    });
  });
});
