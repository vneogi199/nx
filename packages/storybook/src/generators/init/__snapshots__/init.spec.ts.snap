// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/storybook:init dependencies for package.json should add angular related dependencies when using Angular as uiFramework 1`] = `
{
  "dependencies": {
    "existing": "1.0.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
  },
  "devDependencies": {
    "@angular/forms": "*",
    "@nx/js": "0.0.1",
    "@nx/storybook": "0.0.1",
    "@storybook/addon-essentials": "7.2.2",
    "@storybook/angular": "7.2.2",
    "@storybook/core-server": "7.2.2",
    "@swc-node/register": "~1.6.7",
    "@swc/core": "~1.3.85",
    "existing": "1.0.0",
    "prettier": "^2.6.2",
    "typescript": "~5.1.3",
  },
  "name": "@proj/source",
}
`;
